app/src/main/java/org/mozilla/focus/telemetry/GleanMetricsService.kt

```diff
@@ -0,0 +1,104 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.telemetry
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.search.SearchEngine
+import mozilla.components.browser.state.state.selectedOrDefaultSearchEngine
+import mozilla.components.feature.search.ext.waitForSelectedOrDefaultSearchEngine
+import mozilla.components.service.glean.net.ConceptFetchHttpUploader
+import mozilla.telemetry.glean.Glean
+import mozilla.telemetry.glean.config.Configuration
+import org.mozilla.focus.BuildConfig
+import org.mozilla.focus.Components
+import org.mozilla.focus.GleanMetrics.GleanBuildInfo
+import org.mozilla.focus.GleanMetrics.Browser
+import org.mozilla.focus.GleanMetrics.LegacyIds
+import org.mozilla.focus.GleanMetrics.Pings
+import org.mozilla.focus.ext.components
+import org.mozilla.focus.telemetry.TelemetryWrapper.isTelemetryEnabled
+import org.mozilla.focus.utils.Settings
+import java.util.UUID
+
+/**
+ * Glean telemetry service.
+ *
+ * To track events, use Glean's generated bindings directly.
+ */
+class GleanMetricsService(context: Context) : MetricsService {
+
+    private val activationPing = ActivationPing(context)
+
+    /**
+     * Initialization flag to
+     */
+    private var initialized = false
+
+    override fun initialize(context: Context) {
+        if (initialized) {
+            return
+        }
+        initialized = true
+
+        val components = context.components
+        val settings = Settings.getInstance(context)
+        val telemetryEnabled = isTelemetryEnabled(context)
+
+        Glean.initialize(
+            applicationContext = context,
+            uploadEnabled = telemetryEnabled,
+            configuration = Configuration(
+                channel = BuildConfig.FLAVOR,
+                httpClient = ConceptFetchHttpUploader(
+                    client = lazy(LazyThreadSafetyMode.NONE) { components.client },
+                    usePrivateRequest = true
+                )
+            ),
+            buildInfo = GleanBuildInfo.buildInfo
+        )
+
+        Glean.registerPings(Pings)
+
+        // Do this immediately after init.
+        MainScope().launch {
```

Comment:
Should this launch in the `GlobalScope`? Creating a scope and not keeping a reference to it, is kind of a smell: It's not really scoped. What part here needs to run on the main thread? 🤔

Reply:
This was partly based on what happened in https://github.com/mozilla-mobile/fenix/issues/19147 and that I wasn't seeing metrics go through until the metrics `set` calls were done on the main thread, but I'll try this on the GlobalScope and re-test.

Thinking about it again, that's probably a bug if so.

Reply:
Documenting what I found out:

- Using `GlobalScope.launch` works.
- Using `CoroutineScope(Dispatchers.IO).launch` works.
- Using `CoroutineScope.launch(Dispatcher.IO)` works, but failed a couple times (might have be network errors).